File: tslang/src/array/insert-middle/ArrayInsertMiddle.ts
const arrayInsertMiddle = <T>(list: T[], newValue: T): T[] => {
	const middleIndex = Math.ceil(list.length / 2);
	const newList = new Array<T>(list.length + 1);
	newList[middleIndex] = newValue;

	for (const [index, value] of list.entries()) {
		if (index >= middleIndex) {
			newList[index + 1] = value;
		} else {
			newList[index] = value;
		}
	}

	return newList;
};

export { arrayInsertMiddle };

File: ktlang/app/src/main/kotlin/ktlang/array/insert-middle/ArrayInsertMiddle.kt
package ktlang.array.insert.middle

fun <T> arrayInsertMiddle(list: List<T>, newValue: T): List<T> {
    val newList: MutableList<T> = mutableListOf<T>()

    val middleIndex = Math.ceil(list.size.toDouble() / 2).toInt()

    for (i in list.indices) {
        if (i == middleIndex) {
            newList.add(newValue)
            newList.add(list[i])
        } else {
            newList.add(list[i])
        }
    }

    return newList
}
